### PR TITLE
[CGALPlugin] Adds explicit definition of features.

### PR DESCRIPTION
**Pull request by IPeterlik**
ADD: new functionality for mesh generation from image: allows for explicit definition of features, i.e., set of positions which will be maintained as sub-set of the mesh nodes. The code was added to the MeshGenerationFromImage components, an example scene was added together with a test that verifies that all features (specified in a VTK file) become nodes of the new mesh. 

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by damienmarchal** [comment]
Hi Igor,

Thank you a lot for your PR. It seems it build successfully on each machine of the CI. 
I will do a quick review of it... anyone else (@dequidt you like cgal if I rememeber) ?

File: applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp [hunk 1]
```diff
@@ -0,0 +1,222 @@
+/******************************************************************************
```
Comment (damienmarchal) [inline applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp]: Many thank to you for adding tests !!

File: applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+namespace sofa{
+namespace {
+
+using namespace modeling;
+using core::objectmodel::New;
+using sofa::simulation::SceneLoaderXML;
+using sofa::core::ExecParams;
+
+
+//template<typename DataTypes>
```
Comment (damienmarchal) [inline applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp]: What is the purpose of all the commented code ? 
In general every commented code should be removed.

File: applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+        meshGenerator = dynamic_cast<cgal::MeshGenerationFromImage< defaulttype::Vec3dTypes, defaulttype::ImageUC >*>(root->getObject("generator"));
+
+        helper::ReadAccessor< Data<sofa::defaulttype::Vec3dTypes::VecCoord> > features = meshGenerator->d_features;
+        helper::ReadAccessor< Data<sofa::defaulttype::Vec3dTypes::VecCoord> > nodes = meshGenerator->f_newX0;
+
+        double tolerance = 1e-10;
+        size_t nfts = features.size();
+        bool foundFeatures[nfts];
+        std::cout << "#features to be tested: " << nfts << std::endl;
```
Comment (damienmarchal) [inline applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp]: std::cout must be replaced with msg_* API. 
You can also use the gtest << feature do provide details about tests: 
EXPECT_EQ(blabhah) << "Provide here the details of failing test you want t add as the number of feature tested: ";

File: applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                return false;
+        }
+
+        return true;
+
+	}
+
+    void testDataFields()
+    {
+        std::cout << "Calling test fields!" << std::endl;
```
Comment (damienmarchal) [inline applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp]: This code should be removed or completed.

File: applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        EXPECT_EQ(grabber->d_fullFrame.getName(), "fullFrame");
+        EXPECT_EQ(grabber->d_frame1.getName(), "img1");
+        EXPECT_EQ(grabber->d_frame2.getName(), "img2");
+        EXPECT_EQ(grabber->d_fps.getName(), "fps");
+        EXPECT_EQ(grabber->d_dimensions.getName(), "dimensions");*/
+    }
+
+    void SetUp()
+    {
+        /// a scene defined using raw-strings (C++11)
```
Comment (damienmarchal) [inline applications/plugins/CGALPlugin/CGALPlugin_test/MeshFromImageWithFeatures_test.cpp]: +1 for the raw-string...this look super great way to write scene in tests. Thank for that tip.

File: applications/plugins/CGALPlugin/MeshGenerationFromImage.h [hunk 6]
```diff
@@ -146,7 +155,7 @@ class MeshGenerationFromImage : public sofa::core::DataEngine
     sofa::core::objectmodel::Data<bool> frozen;
 
     //Parameters
-    sofa::core::objectmodel::Data<double> facetAngle, facetSize, facetApproximation;
+    sofa::core::objectmodel::Data<double> edgeSize, facetAngle, facetSize, facetApproximation;
```
Comment (damienmarchal) [inline applications/plugins/CGALPlugin/MeshGenerationFromImage.h]: edgeSize should follow the d_edgeSize naming convention...but for consistency let's say it is not a problem here :)

**Review by damienmarchal — COMMENTED** [review]
Thanks a lot for this PR that adds new features and provides examples scenes & tests.

There is minor cosmetic details to fix that I pointed. 


To go further than the scope of this PR (if someone is passing by and have free time):
     - The component should be clean to stick to the sofa style. 
     - Add more tests by playing on the all the parameter to validate more intensively the different code path.

File: applications/plugins/CGALPlugin/MeshGenerationFromImage.inl [hunk 7]
```diff
@@ -293,6 +329,7 @@ void MeshGenerationFromImage<DataTypes, _ImageTypes>::update()
         p[2] = CGAL::to_double(pointCgal.z());
         if (Vnbe.find(vit) == Vnbe.end() || Vnbe[vit] <= 0)
         {
+            std::cout << "Un-connected point: " << p << std::endl;
```
Comment (damienmarchal) [inline applications/plugins/CGALPlugin/MeshGenerationFromImage.inl]: std::cout is not allowed by the Sofa coding guidelines.

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by guparan** [comment]
@IPeterlik, have you seen Damien's review?

**Comment by damienmarchal** [comment]
[ci-build]

**Pull request merged**